File: runtime/node/vssh/trabalhos.js
'use strict';

// vssh.trabalhos: um trabalho longo que o app acompanha, na estação ou na fila, com a mesma cara.
// O par de `vssh/trabalhos.py`, com o mesmo registro em disco, os mesmos eventos e as mesmas regras.
//
//   const { trabalhos } = require('vssh');
//   const t = trabalhos.rodar('transcrever:3f2a', ['escriba-motor', '--entrada', a, '--saida', r],
//     { titulo: 'Entrevista 03.m4a', aoEvento: publicar });
//   const t2 = trabalhos.naFila('transcrever:3f2a', pedido, { titulo, aoEvento, linhas: true });
//   await t.esperar();   // o registro final
//   t.cancelar();
//
// `aoEvento(evento, registro)` recebe `estado`, `progresso`, `linha` (o stdout que não é progresso,
// com o texto em `registro.linha`) e `fim` (concluido, falhou ou cancelado, com `motivo` e
// `codigo`).
//
// Na estação, o processo sobe num grupo próprio, e cancelar manda SIGTERM ao grupo e SIGKILL cinco
// segundos depois; as linhas de progresso viram a atividade na bandeja, e o fim vira notificação
// antes de a atividade sair. Na fila, a biblioteca não escreve atividade nem notificação, porque o
// portal já faz as duas coisas para o job. O registro de cada trabalho mora em
// `$VSSH_APP_DATA_DIR/trabalhos/<chave>.json`, e `retomar()` volta a seguir os da fila depois de um
// reinício e encerra os da estação, que perderam o stdout.

const fs = require('node:fs');
const path = require('node:path');
const { spawn } = require('node:child_process');
const readline = require('node:readline');

const app = require('./app.js');
const avisos = require('./avisos.js');
const fila = require('./fila.js');
const progresso = require('./progresso.js');

const FINAIS = ['concluido', 'falhou', 'cancelado'];
const CHAVE = /^[\w:.-]{1,64}$/;
const ESPERA_DO_SIGKILL_MS = 5000;
const INTERVALO_DA_ATIVIDADE_MS = 500;

const emCurso = new Map();
let pararRenovacao = null;

const diretorio = (env) => path.join(app.dados(env), 'trabalhos');
const arquivo = (chave, env) => path.join(diretorio(env), `${chave}.json`);

function gravar(registro, env) {
  const alvo = arquivo(registro.chave, env);
  try {
    fs.mkdirSync(path.dirname(alvo), { recursive: true });
    fs.writeFileSync(`${alvo}.tmp`, JSON.stringify(registro));
    fs.renameSync(`${alvo}.tmp`, alvo);
  } catch (e) {
    console.warn(`[vssh.trabalhos] não foi possível gravar ${alvo}: ${e.message}`);
  }
}

/** O registro gravado de um trabalho, ou `null`. */
function estado(chave, env = process.env) {
  try {
    const r = JSON.parse(fs.readFileSync(arquivo(chave, env), 'utf8'));
    return r && typeof r === 'object' ? r : null;
  } catch { return null; }
}

/** Os registros gravados, do mais novo ao mais velho. */
function listar(env = process.env) {
  let nomes = [];
  try { nomes = fs.readdirSync(diretorio(env)).filter((n) => n.endsWith('.json')); } catch { return []; }
  return nomes.map((n) => estado(n.slice(0, -5), env)).filter(Boolean)
    .sort((a, b) => (b.criadoEm || 0) - (a.criadoEm || 0));
}

function sinalizarGrupo(pid, sinal) {
  try { process.kill(-pid, sinal); } catch { /* o grupo já acabou */ }
}

class Trabalho {
  constructor(registro, aoEvento, env) {
    this.chave = registro.chave;
    this._reg = registro;
    this._aoEvento = aoEvento;
    this._env = env;
    this._cancelado = false;
    this._processo = null;
    this._ultimaAtividade = 0;
    this._fim = new Promise((resolve) => { this._resolver = resolve; });
    this._terminado = FINAIS.includes(registro.estado);
    if (this._terminado) this._resolver({ ...registro });
  }

  estado() { return { ...this._reg }; }

  /** O registro final. */
  esperar() { return this._fim; }

  /** Pede o fim. Na estação, SIGTERM ao grupo e SIGKILL depois; na fila, o cancelar do portal. */
  cancelar() {
    if (this._terminado) return false;
    this._cancelado = true;
    if (this._reg.onde === 'estacao' && this._processo) {
      const p = this._processo;
      sinalizarGrupo(p.pid, 'SIGTERM');
      setTimeout(() => { if (p.exitCode === null && p.signalCode === null) sinalizarGrupo(p.pid, 'SIGKILL'); }, ESPERA_DO_SIGKILL_MS).unref();
    } else if (this._reg.onde === 'fila' && this._reg.job) {
      fila.cancelar(this._reg.job, this._env).catch(() => {});
    }
    return true;
  }

  _mudar(evento, campos) {
    Object.assign(this._reg, campos);
    const registro = { ...this._reg };
    gravar(registro, this._env);
    if (evento && this._aoEvento) {
      try { this._aoEvento(evento, registro); } catch (e) { console.warn(`[vssh.trabalhos] aoEvento(${evento}) de ${this.chave} falhou: ${e.message}`); }
    }
    return registro;
  }

  _linha(texto) {
    if (!this._aoEvento) return;
    try { this._aoEvento('linha', { ...this._reg, linha: texto }); } catch (e) { console.warn(`[vssh.trabalhos] aoEvento(linha) de ${this.chave} falhou: ${e.message}`); }
  }

  // Grava o fim, faz o que precisa estar feito antes de alguém saber dele (a notificação e a
  // atividade), e só então avisa o app e solta quem espera. Na ordem inversa, um backend que
  // encerra logo depois de `esperar()` perderia a notificação.
  _terminar(estadoFinal, motivo = null, codigo = null, antesDeAnunciar = null) {
    if (emCurso.get(this.chave) === this) emCurso.delete(this.chave);
    this._terminado = true;
    const registro = this._mudar(null, { estado: estadoFinal, motivo, codigo, terminadoEm: Date.now() });
    if (antesDeAnunciar) {
      try { antesDeAnunciar(registro); } catch (e) { console.warn(`[vssh.trabalhos] o fim de ${this.chave} não foi registrado: ${e.message}`); }
    }
    if (this._aoEvento) {
      try { this._aoEvento('fim', { ...registro }); } catch (e) { console.warn(`[vssh.trabalhos] aoEvento(fim) de ${this.chave} falhou: ${e.message}`); }
    }
    this._resolver(registro);
    return registro;
  }
}

function registrar(chave, onde, titulo, extra) {
  if (typeof chave !== 'string' || !CHAVE.test(chave)) {
    throw new Error(`chave de trabalho inválida (use letras, números, ":", "." e "-", até 64): ${chave}`);
  }
  const vivo = emCurso.get(chave);
  if (vivo && !vivo._terminado) throw new Error(`já há um trabalho em curso com a chave ${chave}`);
  return {
    chave, titulo: titulo || chave, onde, estado: onde === 'fila' ? 'enviando' : 'rodando',
    progresso: null, motivo: null, codigo: null, criadoEm: Date.now(), terminadoEm: null, ...extra,
  };
}

function itemDaAtividade(reg) {
  const p = reg.progresso || {};
  return {
    titulo: reg.titulo,
    texto: p.etapa || 'Rodando',
    formato: 'progresso',
    progresso: p.total ? { feito: p.feito, total: p.total } : { indeterminado: true },
  };
}

function notificarFim(reg, opcaoRegistrar, acoes, rota, abrir, env) {
  if (opcaoRegistrar === false || reg.estado === 'cancelado') return;   // quem cancelou já sabe
  let corpo;
  if (typeof opcaoRegistrar === 'function') corpo = opcaoRegistrar(reg);
  else if (reg.estado === 'concluido') corpo = { titulo: reg.titulo, texto: 'Terminou.', level: 'success' };
  else corpo = { titulo: reg.titulo, texto: reg.motivo || 'Falhou.', level: 'error' };
  if (!corpo) return;
  avisos.notificar(corpo.texto || '', {
    titulo: corpo.titulo, nivel: corpo.level, chave: `trabalho:${reg.chave}:${reg.criadoEm}`, acoes, rota, abrir, env,
  });
}

/**
 * Sobe `comando` (um array) e o acompanha. Devolve o `Trabalho`.
 *
 * Opções: `titulo`, `aoEvento`, `cwd`, `ambiente` (o do processo; o do app por padrão),
 * `registrar` (`true`, `false` ou `(registro) => ({ titulo, texto, level })`), `acoes`, `rota` e
 * `abrir` (os de `avisos.notificar`: botões na notificação, e o caminho dentro do app aonde o
 * clique nela leva), `env`.
 */
function rodar(chave, comando, opcoes = {}) {
  const env = opcoes.env || process.env;
  const t = new Trabalho(registrar(chave, 'estacao', opcoes.titulo, { pid: null }), opcoes.aoEvento, env);
  const [exe, ...args] = comando;
  const processo = spawn(exe, args, {
    cwd: opcoes.cwd, env: opcoes.ambiente || process.env, stdio: ['ignore', 'pipe', 'pipe'], detached: true,
  });
  t._processo = processo;
  emCurso.set(chave, t);
  t._mudar(null, { pid: processo.pid ?? null });
  if (!pararRenovacao) pararRenovacao = avisos.manterAtividadesVivas(env);
  avisos.atividade(chave, itemDaAtividade(t._reg), env);

  const erros = [];
  readline.createInterface({ input: processo.stderr }).on('line', (linha) => {
    if (linha.trim()) { erros.push(linha); if (erros.length > 20) erros.shift(); }
  });
  const saida = readline.createInterface({ input: processo.stdout });
  saida.on('line', (linha) => {
    const p = progresso.ler(linha);
    if (!p) { t._linha(linha); return; }
    const registro = t._mudar('progresso', { progresso: p });
    const agora = Date.now();
    if (agora - t._ultimaAtividade >= INTERVALO_DA_ATIVIDADE_MS) {
      t._ultimaAtividade = agora;
      avisos.atividade(chave, itemDaAtividade(registro), env);
    }
  });

  let saidaFechou = new Promise((r) => saida.on('close', r));
  let terminou = false;
  // Um `spawn` que falha emite `error` e depois `close`; o trabalho termina uma vez só.
  const terminar = (codigo, motivoDeSpawn) => {
    if (terminou) return;
    terminou = true;
    saidaFechou.then(() => {
      let estadoFinal = 'concluido';
      let motivo = null;
      if (t._cancelado) estadoFinal = 'cancelado';
      else if (motivoDeSpawn) { estadoFinal = 'falhou'; motivo = motivoDeSpawn; }
      else if (codigo !== 0) { estadoFinal = 'falhou'; motivo = erros.at(-1) || `saiu com o código ${codigo}`; }
      t._terminar(estadoFinal, motivo, codigo, (registro) => {
        notificarFim(registro, opcoes.registrar ?? true, opcoes.acoes, opcoes.rota, opcoes.abrir, env);
        avisos.limparAtividade(chave, { env });
      });
    });
  };
  processo.on('error', (e) => { saidaFechou = Promise.resolve(); terminar(null, `não consegui rodar ${exe}: ${e.message}`); });
  processo.on('close', (codigo) => terminar(codigo));
  return t;
}

/** Submete `pedido` à fila (o formato de `fila.submeter`) e o acompanha. Devolve o `Trabalho`. */
function naFila(chave, pedido, opcoes = {}) {
  const env = opcoes.env || process.env;
  const t = new Trabalho(registrar(chave, 'fila', opcoes.titulo, { job: null }), opcoes.aoEvento, env);
  emCurso.set(chave, t);
  t._mudar('estado', {});
  (async () => {
    let id;
    try { id = await fila.submeter(pedido, undefined, env); } catch (e) { t._terminar('falhou', e.mensagem || e.message); return; }
    t._mudar('estado', { job: id, estado: 'enviado' });
    if (t._cancelado) t.cancelar();
    await seguir(t, !!opcoes.linhas, env);
  })();
  return t;
}

async function seguir(t, linhas, env) {
  const id = t._reg.job;
  const vistas = new Set();
  let seguindoLog = false;

  const seguirLog = async () => {
    // O log reabre se o stream cair com o job ainda rodando, e o portal manda de novo as últimas
    // linhas: as já entregues não saem duas vezes.
    while (!t._terminado) {
      try {
        for await (const linha of fila.log(id, env)) {
          if (vistas.has(linha) || progresso.ler(linha)) continue;
          vistas.add(linha);
          if (vistas.size > 2000) vistas.delete(vistas.values().next().value);
          t._linha(linha);
        }
      } catch { /* o stream caiu: tenta de novo enquanto o job roda */ }
      if (t._terminado) return;
      await new Promise((r) => setTimeout(r, 3000).unref());
    }
  };

  const aoEvento = (evento, job) => {
    if (evento === 'estado') {
      t._mudar('estado', { estado: job.estado || t._reg.estado, motivo: job.motivo ?? null });
      if (linhas && job.estado === 'rodando' && !seguindoLog) { seguindoLog = true; seguirLog(); }
    } else if (evento === 'progresso' && job.progresso) {
      const p = job.progresso;
      t._mudar('progresso', { progresso: { feito: p.feito, total: p.total ?? null, etapa: p.etapa ?? null } });
    }
  };

  let final;
  try { final = await fila.acompanhar(id, aoEvento, env); } catch (e) { t._terminar('falhou', e.mensagem || e.message); return; }
  t._terminar(final.estado || 'falhou', final.motivo ?? null, final.exit ?? null);
}

function vivo(pid) {
  try { process.kill(pid, 0); return true; } catch { return false; }
}

/**
 * Volta a seguir os trabalhos da fila que estavam em curso, e encerra os da estação. Devolve os
 * `Trabalho` retomados. Chame no boot do backend, antes de servir.
 */
function retomar(opcoes = {}) {
  const env = opcoes.env || process.env;
  const retomados = [];
  for (const reg of listar(env)) {
    if (FINAIS.includes(reg.estado) || emCurso.has(reg.chave)) continue;
    const t = new Trabalho(reg, opcoes.aoEvento, env);
    if (reg.onde === 'fila' && reg.job) {
      emCurso.set(t.chave, t);
      seguir(t, !!opcoes.linhas, env);
      retomados.push(t);
    } else {
      if (reg.pid && vivo(reg.pid)) sinalizarGrupo(reg.pid, 'SIGKILL');
      t._terminar('falhou', reg.onde === 'estacao' ? 'o app reiniciou no meio' : 'o envio à fila foi interrompido', null,
        () => avisos.limparAtividade(t.chave, { env }));
    }
  }
  return retomados;
}

module.exports = { rodar, naFila, estado, listar, retomar, Trabalho, FINAIS };

File: runtime/python/vssh/trabalhos.py
"""
vssh.trabalhos: um trabalho longo que o app acompanha, na estação ou na fila, com a mesma cara.

    from vssh import trabalhos

    t = trabalhos.rodar('transcrever:3f2a', ['escriba-motor', '--entrada', a, '--saida', r],
                        titulo='Entrevista 03.m4a', ao_evento=publicar)
    t = trabalhos.na_fila('transcrever:3f2a', pedido, titulo='Entrevista 03.m4a',
                          ao_evento=publicar, linhas=True)

    t.estado()          # o registro: estado, progresso, motivo, código, job da fila
    t.cancelar()
    t.esperar(timeout)  # o registro final, ou o atual se o prazo acabar antes

`ao_evento(evento, registro)` recebe:

    estado      o trabalho mudou de estado (enviando, enviado, na_fila, rodando)
    progresso   o trabalho informou um progresso novo (`vssh.progresso`)
    linha       uma linha do stdout que não é de progresso, sem o `\\n`: o protocolo do próprio app
    fim         concluido, falhou ou cancelado; `motivo` e `codigo` dizem por quê

─── Na estação ──────────────────────────────────────────────────────────────

`rodar` sobe o processo num grupo próprio, e cancelar manda SIGTERM ao grupo inteiro: um motor
que abriu um `ffmpeg` filho não o deixa rodando sozinho. Cinco segundos depois, quem ainda estiver
de pé recebe SIGKILL. O stdout é lido linha a linha; as de progresso viram a atividade do trabalho
na bandeja, e as outras vão para o app. As últimas linhas do stderr dão o `motivo` de uma falha.
No fim, a notificação vai ao sino antes de a atividade sair, como em `avisos.limpar_atividade`.

─── Na fila ─────────────────────────────────────────────────────────────────

`na_fila` submete o pedido e segue o job pelos eventos do portal. Aqui a biblioteca não escreve
atividade nem notificação: o portal já põe o job em curso na bandeja e avisa o fim, e escrever de
novo daria duas linhas para o mesmo trabalho. Com `linhas=True`, ela segue também o log do
container quando ele começa a rodar, e entrega como `linha` o que não é progresso.

─── Depois de um reinício ──────────────────────────────────────────────────

O registro de cada trabalho mora em `$VSSH_APP_DATA_DIR/trabalhos/<chave>.json`, reescrito a cada
mudança. `retomar()` volta a seguir os da fila que estavam em curso, porque o job continua no
cluster. Um da estação não tem volta: o processo não é mais filho de ninguém, e o stdout dele se
perdeu. Ele é encerrado, para não segurar a GPU, e marcado como `falhou` com o motivo.

A chave segue a regra das atividades do shell (`[\\w:.-]`, até 64 caracteres), porque ela é a
chave da atividade na bandeja. Uma chave só tem um trabalho em curso de cada vez.
"""

import collections
import json
import os
import re
import signal
import subprocess
import threading
import time

from . import app as _app
from . import avisos
from . import fila
from . import progresso as _progresso

__all__ = ['rodar', 'na_fila', 'estado', 'listar', 'retomar', 'Trabalho', 'ESTADOS_FINAIS']

ESTADOS_FINAIS = ('concluido', 'falhou', 'cancelado')
_CHAVE = re.compile(r'^[\w:.-]{1,64}$')
_ESPERA_DO_SIGKILL_S = 5.0
# A bandeja não precisa de mais que duas atualizações por segundo, e um motor que informa cada
# milésimo escreveria o arquivo da atividade mil vezes.
_INTERVALO_DA_ATIVIDADE_S = 0.5

_tranca = threading.Lock()
_em_curso = {}
_renovacao = []


def _agora_ms():
    return int(time.time() * 1000)


def _diretorio(env):
    return os.path.join(_app.dados(env), 'trabalhos')


def _arquivo(chave, env):
    return os.path.join(_diretorio(env), chave + '.json')


def estado(chave, env=None):
    """O registro gravado de um trabalho, ou `None`."""
    try:
        with open(_arquivo(chave, env), encoding='utf-8') as f:
            r = json.load(f)
        return r if isinstance(r, dict) else None
    except (OSError, ValueError):
        return None


def listar(env=None):
    """Os registros gravados, do mais novo ao mais velho."""
    try:
        nomes = [n for n in os.listdir(_diretorio(env)) if n.endswith('.json')]
    except OSError:
        return []
    registros = [estado(n[:-5], env) for n in nomes]
    return sorted((r for r in registros if r), key=lambda r: r.get('criadoEm') or 0, reverse=True)


class Trabalho:
    """Um trabalho em curso ou terminado. Quem cria é `rodar`, `na_fila` ou `retomar`."""

    def __init__(self, registro, ao_evento, env):
        self.chave = registro['chave']
        self._reg = registro
        self._ao_evento = ao_evento
        self._env = env
        self._tranca = threading.Lock()
        self._fim = threading.Event()
        self._cancelado = False
        self._processo = None
        self._ultima_atividade = 0.0
        if registro['estado'] in ESTADOS_FINAIS:
            self._fim.set()

    def estado(self):
        with self._tranca:
            return dict(self._reg)

    def esperar(self, timeout=None):
        self._fim.wait(timeout)
        return self.estado()

    def cancelar(self):
        """Pede o fim. Na estação, SIGTERM ao grupo e SIGKILL depois; na fila, o cancelar do portal."""
        if self._fim.is_set():
            return False
        self._cancelado = True
        if self._reg['onde'] == 'estacao' and self._processo is not None:
            _sinalizar_grupo(self._processo.pid, signal.SIGTERM)
            p = self._processo

            def matar():
                if p.poll() is None:
                    _sinalizar_grupo(p.pid, signal.SIGKILL)
            t = threading.Timer(_ESPERA_DO_SIGKILL_S, matar)
            t.daemon = True
            t.start()
        elif self._reg['onde'] == 'fila' and self._reg.get('job'):
            try:
                fila.cancelar(self._reg['job'], env=self._env)
            except fila.ErroDaFila:
                pass
        return True

    # ── por dentro ──────────────────────────────────────────────────────────

    def _mudar(self, evento, **campos):
        with self._tranca:
            self._reg.update(campos)
            registro = dict(self._reg)
            _gravar(registro, self._env)
        if evento and self._ao_evento:
            try:
                self._ao_evento(evento, registro)
            except Exception as e:  # o app que ouve não derruba o trabalho
                print('[vssh.trabalhos] ao_evento(%s) de %s falhou: %s' % (evento, self.chave, e))
        return registro

    def _linha(self, texto):
        if self._ao_evento:
            try:
                self._ao_evento('linha', dict(self._reg, linha=texto))
            except Exception as e:
                print('[vssh.trabalhos] ao_evento(linha) de %s falhou: %s' % (self.chave, e))

    def _terminar(self, estado_final, motivo=None, codigo=None, antes_de_anunciar=None):
        """Grava o fim, faz o que precisa estar feito antes de alguém saber dele (a notificação e a
        atividade), e só então avisa o app e solta quem espera. Na ordem inversa, um backend que
        encerra logo depois de `esperar()` perderia a notificação."""
        with _tranca:
            if _em_curso.get(self.chave) is self:
                del _em_curso[self.chave]
        registro = self._mudar(None, estado=estado_final, motivo=motivo, codigo=codigo, terminadoEm=_agora_ms())
        if antes_de_anunciar:
            try:
                antes_de_anunciar(registro)
            except Exception as e:
                print('[vssh.trabalhos] o fim de %s não foi registrado: %s' % (self.chave, e))
        if self._ao_evento:
            try:
                self._ao_evento('fim', dict(registro))
            except Exception as e:
                print('[vssh.trabalhos] ao_evento(fim) de %s falhou: %s' % (self.chave, e))
        self._fim.set()
        return registro


def _gravar(registro, env):
    arquivo = _arquivo(registro['chave'], env)
    try:
        os.makedirs(os.path.dirname(arquivo), exist_ok=True)
        tmp = arquivo + '.tmp'
        with open(tmp, 'w', encoding='utf-8') as f:
            json.dump(registro, f, ensure_ascii=False, separators=(',', ':'))
        os.replace(tmp, arquivo)
    except OSError as e:
        print('[vssh.trabalhos] não foi possível gravar %s: %s' % (arquivo, e))


def _sinalizar_grupo(pid, sinal):
    try:
        os.killpg(pid, sinal)
    except (ProcessLookupError, PermissionError):
        pass


def _registrar(chave, onde, titulo, env, **extra):
    if not isinstance(chave, str) or not _CHAVE.match(chave):
        raise ValueError('chave de trabalho inválida (use letras, números, ":", "." e "-", até 64): %r' % (chave,))
    with _tranca:
        vivo = _em_curso.get(chave)
        if vivo is not None and not vivo._fim.is_set():
            raise ValueError('já há um trabalho em curso com a chave %s' % chave)
    return dict({'chave': chave, 'titulo': titulo or chave, 'onde': onde, 'estado': 'enviando' if onde == 'fila' else 'rodando',
                 'progresso': None, 'motivo': None, 'codigo': None, 'criadoEm': _agora_ms(), 'terminadoEm': None}, **extra)


def _acompanhar_bandeja(env):
    with _tranca:
        if not _renovacao:
            _renovacao.append(avisos.manter_atividades_vivas(env=env))


# ── Na estação ────────────────────────────────────────────────────────────────

def _item_da_atividade(reg):
    p = reg.get('progresso') or {}
    return {
        'titulo': reg['titulo'],
        'texto': p.get('etapa') or 'Rodando',
        'formato': 'progresso',
        'progresso': {'feito': p['feito'], 'total': p['total']} if p.get('total') else {'indeterminado': True},
    }


def rodar(chave, comando, titulo=None, ao_evento=None, cwd=None, ambiente=None, registrar=True,
          acoes=None, rota=None, abrir=None, env=None):
    """Sobe `comando` (uma lista) e o acompanha. Devolve o `Trabalho`.

    `ambiente` é o ambiente do processo (o do app, por padrão). `registrar` decide a notificação
    do fim: `True` escreve uma com o título e o resultado, `False` não escreve, e uma função
    `registrar(registro) -> {titulo, texto, level} | None` escreve a que ela devolver. `acoes`,
    `rota` e `abrir` são os de `avisos.notificar`: botões na notificação, e o caminho dentro do
    app aonde o clique nela leva.
    """
    env = os.environ if env is None else env
    reg = _registrar(chave, 'estacao', titulo, env, pid=None)
    t = Trabalho(reg, ao_evento, env)
    processo = subprocess.Popen(list(comando), cwd=cwd, env=dict(ambiente) if ambiente is not None else None,
                                stdin=subprocess.DEVNULL, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
                                text=True, bufsize=1, start_new_session=True)
    t._processo = processo
    with _tranca:
        _em_curso[chave] = t
    t._mudar(None, pid=processo.pid)
    _acompanhar_bandeja(env)
    avisos.atividade(chave, _item_da_atividade(reg), env=env)

    erros = collections.deque(maxlen=20)

    def ler_stderr():
        for linha in processo.stderr:
            if linha.strip():
                erros.append(linha.rstrip('\r\n'))

    def ler_stdout():
        for linha in processo.stdout:
            p = _progresso.ler(linha)
            if p is None:
                t._linha(linha.rstrip('\r\n'))
                continue
            registro = t._mudar('progresso', progresso=p)
            agora = time.monotonic()
            if agora - t._ultima_atividade >= _INTERVALO_DA_ATIVIDADE_S:
                t._ultima_atividade = agora
                avisos.atividade(chave, _item_da_atividade(registro), env=env)

    def esperar():
        leitores = [threading.Thread(target=ler_stdout, daemon=True), threading.Thread(target=ler_stderr, daemon=True)]
        for l in leitores:
            l.start()
        codigo = processo.wait()
        for l in leitores:
            l.join()
        if t._cancelado:
            estado_final, motivo = 'cancelado', None
        elif codigo == 0:
            estado_final, motivo = 'concluido', None
        else:
            estado_final, motivo = 'falhou', (erros[-1] if erros else 'saiu com o código %d' % codigo)
        def registrar_fim(registro):
            _notificar_fim(registro, registrar, acoes, rota, abrir, env)
            avisos.limpar_atividade(chave, env=env)
        t._terminar(estado_final, motivo=motivo, codigo=codigo, antes_de_anunciar=registrar_fim)

    threading.Thread(target=esperar, name='vssh-trabalho-%s' % chave, daemon=True).start()
    return t


def _notificar_fim(reg, registrar, acoes, rota, abrir, env):
    if not registrar or reg['estado'] == 'cancelado':   # quem cancelou já sabe
        return
    if callable(registrar):
        corpo = registrar(reg)
    elif reg['estado'] == 'concluido':
        corpo = {'titulo': reg['titulo'], 'texto': 'Terminou.', 'level': 'success'}
    else:
        corpo = {'titulo': reg['titulo'], 'texto': reg.get('motivo') or 'Falhou.', 'level': 'error'}
    if not corpo:
        return
    avisos.notificar(corpo.get('texto') or '', titulo=corpo.get('titulo'), nivel=corpo.get('level'),
                     chave='trabalho:%s:%s' % (reg['chave'], reg['criadoEm']), acoes=acoes, rota=rota, abrir=abrir, env=env)


# ── Na fila ───────────────────────────────────────────────────────────────────

def na_fila(chave, pedido, titulo=None, ao_evento=None, linhas=False, env=None):
    """Submete `pedido` à fila (o formato de `fila.submeter`) e o acompanha. Devolve o `Trabalho`."""
    env = os.environ if env is None else env
    reg = _registrar(chave, 'fila', titulo, env, job=None)
    t = Trabalho(reg, ao_evento, env)
    with _tranca:
        _em_curso[chave] = t
    t._mudar('estado')

    def submeter():
        try:
            ident = fila.submeter(pedido, env=env)
        except fila.ErroDaFila as e:
            t._terminar('falhou', motivo=e.mensagem, codigo=None)
            return
        t._mudar('estado', job=ident, estado='enviado')
        if t._cancelado:
            t.cancelar()
        _seguir(t, linhas, env)

    threading.Thread(target=submeter, name='vssh-trabalho-%s' % chave, daemon=True).start()
    return t


def _seguir(t, linhas, env):
    ident = t._reg['job']
    vistas = collections.OrderedDict()
    seguindo_log = []

    def seguir_log():
        # O log reabre se o stream cair com o job ainda rodando, e o portal manda de novo as
        # últimas linhas: as já entregues não saem duas vezes.
        while not t._fim.is_set():
            try:
                for linha in fila.log(ident, env=env):
                    if linha in vistas or _progresso.ler(linha) is not None:
                        continue
                    vistas[linha] = True
                    if len(vistas) > 2000:
                        vistas.popitem(last=False)
                    t._linha(linha)
            except fila.ErroDaFila:
                pass
            if t._fim.wait(3.0):
                return

    def ao_evento(evento, job):
        if evento == 'estado':
            t._mudar('estado', estado=job.get('estado') or t._reg['estado'], motivo=job.get('motivo'))
            if linhas and job.get('estado') == 'rodando' and not seguindo_log:
                seguindo_log.append(True)
                threading.Thread(target=seguir_log, daemon=True).start()
        elif evento == 'progresso' and job.get('progresso'):
            p = job['progresso']
            t._mudar('progresso', progresso={'feito': p.get('feito'), 'total': p.get('total'), 'etapa': p.get('etapa')})

    try:
        final = fila.acompanhar(ident, ao_evento=ao_evento, env=env)
    except fila.ErroDaFila as e:
        t._terminar('falhou', motivo=e.mensagem)
        return
    t._terminar(final.get('estado') or 'falhou', motivo=final.get('motivo'), codigo=final.get('exit'))


# ── Depois de um reinício ─────────────────────────────────────────────────────

def _vivo(pid):
    try:
        os.kill(pid, 0)
        return True
    except (ProcessLookupError, PermissionError, TypeError):
        return False


def retomar(ao_evento=None, linhas=False, env=None):
    """Volta a seguir os trabalhos da fila que estavam em curso, e encerra os da estação.

    Devolve os `Trabalho` retomados. Chame no boot do backend, antes de servir.
    """
    env = os.environ if env is None else env
    retomados = []
    for reg in listar(env):
        if reg.get('estado') in ESTADOS_FINAIS:
            continue
        with _tranca:
            if reg['chave'] in _em_curso:
                continue
        t = Trabalho(reg, ao_evento, env)
        if reg.get('onde') == 'fila' and reg.get('job'):
            with _tranca:
                _em_curso[t.chave] = t
            threading.Thread(target=_seguir, args=(t, linhas, env), daemon=True).start()
            retomados.append(t)
        else:
            if reg.get('pid') and _vivo(reg['pid']):
                _sinalizar_grupo(reg['pid'], signal.SIGKILL)
            motivo = 'o app reiniciou no meio' if reg.get('onde') == 'estacao' else 'o envio à fila foi interrompido'
            t._terminar('falhou', motivo=motivo, antes_de_anunciar=lambda _r: avisos.limpar_atividade(t.chave, env=env))
    return retomados
